File: python/src/epostak/token_manager.py
"""Thread-safe OAuth token manager for the ePostak SDK.

Automatically mints a JWT via ``POST /sapi/v1/auth/token`` on first use,
caches it, and refreshes 60 seconds before expiry via ``/sapi/v1/auth/renew``.
"""

from __future__ import annotations

import re
import threading
import time
from typing import Optional

import httpx

_SAPI_TOKEN_PATH = "/sapi/v1/auth/token"
_SAPI_RENEW_PATH = "/sapi/v1/auth/renew"
_REFRESH_BUFFER_S = 60


class TokenManager:
    """Manages OAuth ``client_credentials`` JWT lifecycle."""

    def __init__(
        self,
        client_id: str,
        client_secret: str,
        base_url: str,
        firm_id: Optional[str] = None,
    ) -> None:
        self._client_id = client_id
        self._client_secret = client_secret
        self._base_url = base_url
        # firm_id accepted for backward compatibility but not used for token minting

        self._access_token: Optional[str] = None
        self._refresh_token: Optional[str] = None
        self._expires_at: float = 0.0
        self._lock = threading.Lock()

    def _sapi_base_url(self) -> str:
        return re.sub(r"/api/v1/?$", "", self._base_url)

    def get_access_token(self) -> str:
        """Return a valid JWT, minting or refreshing as needed. Thread-safe."""
        if self._access_token and time.time() < self._expires_at - _REFRESH_BUFFER_S:
            return self._access_token

        with self._lock:
            # Double-check after acquiring lock
            if self._access_token and time.time() < self._expires_at - _REFRESH_BUFFER_S:
                return self._access_token
            self._refresh_or_mint()
            return self._access_token  # type: ignore[return-value]

    def _refresh_or_mint(self) -> None:
        if self._refresh_token and self._access_token:
            try:
                self._do_renew()
                return
            except Exception:
                pass  # Fall through to full mint
        self._do_mint()

    def _do_mint(self) -> None:
        url = f"{self._sapi_base_url()}{_SAPI_TOKEN_PATH}"
        headers: dict[str, str] = {"Content-Type": "application/json"}

        body: dict[str, str] = {
            "grant_type": "client_credentials",
            "client_id": self._client_id,
            "client_secret": self._client_secret,
        }

        response = httpx.post(url, json=body, headers=headers, timeout=30.0)
        if not response.is_success:
            text = response.text or response.reason_phrase or "token mint failed"
            raise RuntimeError(f"Token mint failed ({response.status_code}): {text}")

        self._apply(response.json())

    def _do_renew(self) -> None:
        url = f"{self._sapi_base_url()}{_SAPI_RENEW_PATH}"
        response = httpx.post(
            url,
            json={
                "grant_type": "refresh_token",
                "refresh_token": self._refresh_token,
            },
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {self._access_token}",
            },
            timeout=30.0,
        )
        if not response.is_success:
            raise RuntimeError(f"Token renew failed ({response.status_code})")

        self._apply(response.json())

    def _apply(self, data: dict) -> None:
        self._access_token = data["access_token"]
        self._refresh_token = data.get("refresh_token")
        self._expires_at = time.time() + data["expires_in"]

File: dotnet/src/EPostak/TokenManager.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using EPostak.Models;

namespace EPostak;

/// <summary>
/// Manages OAuth JWT tokens for API authentication. Automatically mints tokens
/// via <c>POST /sapi/v1/auth/token</c> and refreshes them via
/// <c>POST /sapi/v1/auth/renew</c> 60 seconds before expiry. Thread-safe.
/// </summary>
internal sealed class TokenManager
{
    private const string SapiTokenPath = "/sapi/v1/auth/token";
    private const string SapiRenewPath = "/sapi/v1/auth/renew";
    private static readonly TimeSpan RefreshBuffer = TimeSpan.FromSeconds(60);

    private readonly HttpClient _http;
    private readonly string _clientId;
    private readonly string _clientSecret;
    private readonly string _sapiBaseUrl;

    private readonly SemaphoreSlim _semaphore = new(1, 1);
    private string? _accessToken;
    private string? _refreshToken;
    private DateTimeOffset _expiresAt = DateTimeOffset.MinValue;

    internal TokenManager(HttpClient http, string clientId, string clientSecret, string baseUrl, string? firmId = null)
    {
        _http = http;
        _clientId = clientId;
        _clientSecret = clientSecret;
        // firmId accepted for backward compat but not used for token minting

        // Strip /api/v1 suffix to get the origin for SAPI endpoints
        var idx = baseUrl.IndexOf("/api/v1", StringComparison.Ordinal);
        _sapiBaseUrl = idx >= 0 ? baseUrl[..idx] : baseUrl.TrimEnd('/');
    }

    /// <summary>
    /// Get a valid access token, minting or refreshing as needed.
    /// </summary>
    internal async Task<string> GetAccessTokenAsync(CancellationToken ct = default)
    {
        // Fast path: token still valid
        if (_accessToken is not null && DateTimeOffset.UtcNow < _expiresAt - RefreshBuffer)
            return _accessToken;

        await _semaphore.WaitAsync(ct).ConfigureAwait(false);
        try
        {
            // Double-check after acquiring lock
            if (_accessToken is not null && DateTimeOffset.UtcNow < _expiresAt - RefreshBuffer)
                return _accessToken;

            await RefreshOrMintAsync(ct).ConfigureAwait(false);
            return _accessToken!;
        }
        finally
        {
            _semaphore.Release();
        }
    }

    private async Task RefreshOrMintAsync(CancellationToken ct)
    {
        if (_refreshToken is not null && _accessToken is not null)
        {
            try
            {
                await DoRenewAsync(ct).ConfigureAwait(false);
                return;
            }
            catch
            {
                // Refresh failed — fall through to full mint
            }
        }

        await DoMintAsync(ct).ConfigureAwait(false);
    }

    private async Task DoMintAsync(CancellationToken ct)
    {
        var url = $"{_sapiBaseUrl}{SapiTokenPath}";
        using var request = new HttpRequestMessage(HttpMethod.Post, url);

        var body = new Dictionary<string, string>
        {
            ["grant_type"] = "client_credentials",
            ["client_id"] = _clientId,
            ["client_secret"] = _clientSecret,
        };

        var json = JsonSerializer.Serialize(body, HttpRequestor.JsonOptions);
        request.Content = new StringContent(json, Encoding.UTF8, "application/json");

        HttpResponseMessage response;
        try
        {
            response = await _http.SendAsync(request, ct).ConfigureAwait(false);
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException)
        {
            throw new EPostakException($"Token mint network error: {ex.Message}", ex);
        }

        using (response)
        {
            if (!response.IsSuccessStatusCode)
            {
                var errBody = await response.Content.ReadAsStringAsync(ct).ConfigureAwait(false);
                throw new EPostakException((int)response.StatusCode, $"Token mint failed: {errBody}");
            }

            var stream = await response.Content.ReadAsStreamAsync(ct).ConfigureAwait(false);
            var data = (await JsonSerializer.DeserializeAsync<TokenResponse>(stream, HttpRequestor.JsonOptions, ct).ConfigureAwait(false))!;
            ApplyTokenResponse(data);
        }
    }

    private async Task DoRenewAsync(CancellationToken ct)
    {
        var url = $"{_sapiBaseUrl}{SapiRenewPath}";
        using var request = new HttpRequestMessage(HttpMethod.Post, url);
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _accessToken);

        var body = new Dictionary<string, string>
        {
            ["grant_type"] = "refresh_token",
            ["refresh_token"] = _refreshToken!,
        };

        var json = JsonSerializer.Serialize(body, HttpRequestor.JsonOptions);
        request.Content = new StringContent(json, Encoding.UTF8, "application/json");

        HttpResponseMessage response;
        try
        {
            response = await _http.SendAsync(request, ct).ConfigureAwait(false);
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException)
        {
            throw new EPostakException($"Token renew network error: {ex.Message}", ex);
        }

        using (response)
        {
            if (!response.IsSuccessStatusCode)
                throw new EPostakException((int)response.StatusCode, "Token renew failed");

            var stream = await response.Content.ReadAsStreamAsync(ct).ConfigureAwait(false);
            var data = (await JsonSerializer.DeserializeAsync<TokenResponse>(stream, HttpRequestor.JsonOptions, ct).ConfigureAwait(false))!;
            ApplyTokenResponse(data);
        }
    }

    private void ApplyTokenResponse(TokenResponse data)
    {
        _accessToken = data.AccessToken;
        _refreshToken = data.RefreshToken;
        _expiresAt = DateTimeOffset.UtcNow.AddSeconds(data.ExpiresIn);
    }
}
